## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [1]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [2]:
# Write your code for importing the csv file here
movies = pd.read_csv('Movies.csv')
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moore   
1                       563.0                  1000.0       Orlando Bloom   
2                         0.0                   161.0        Rory Kinnear   
3                     22000.0                 23000.0      Christian Bale   
4                       475.0                   530.0     Samantha Morton   
...                       ...                     ...                 ...   
3848                    291.0                     8.0      David Sullivan   
3849                      0.0                     0.0     Edgar Tancangco   
3850                      0.0                     6.0     Peter Marquardt   
3851                      0.0                   133.0  Caitlin FitzGerald   
3852                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes        gross  \
0                     1000.0  760505847.0   
1                    40000.0  309404152.0   
2                    11000.0  200074175.0   
3                    27000.0  448130642.0   
4                      640.0   73058679.0   
...                      ...          ...   
3848                   291.0     424760.0   
3849                     0.0      70071.0   
3850                   121.0    2040920.0   
3851                   296.0       4584.0   
3852                    86.0      85222.0   

                                   genres  ... num_user_for_reviews language  \
0         Action|Adventure|Fantasy|Sci-Fi  ...               3054.0  English   
1                Action|Adventure|Fantasy  ...               1238.0  English   
2               Action|Adventure|Thriller  ...                994.0  English   
3                         Action|Thriller  ...               2701.0  English   
4                 Action|Adventure|Sci-Fi  ...                738.0  English   
...                                   ...  ...                  ...      ...   
3848                Drama|Sci-Fi|Thriller  ...                371.0  English   
3849                             Thriller  ...                 35.0  English   
3850  Action|Crime|Drama|Romance|Thriller  ...                130.0  Spanish   
3851                         Comedy|Drama  ...                 14.0  English   
3852                          Documentary  ...                 84.0  English   

          country  content_rating       budget  title_year  \
0             USA           PG-13  237000000.0      2009.0   
1             USA           PG-13  300000000.0      2007.0   
2              UK           PG-13  245000000.0      2015.0   
3             USA           PG-13  250000000.0      2012.0   
4             USA           PG-13  263700000.0      2012.0   
...           ...             ...          ...         ...   
3848          USA           PG-13       7000.0      2004.0   
3849  Philippines       Not Rated       7000.0      2005.0   
3850          USA               R       7000.0      1992.0   
3851          USA       Not Rated       9000.0      2011.0   
3852          USA              PG       1100.0      2004.0   

     actor_2_face

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [3]:
# Write your code for inspection here
movies.shape

(3853, 28)

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

In [7]:
print((movies.isnull().sum()>0).sum())

12


### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [8]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations

movies.drop(['color','director_facebook_likes','actor_1_facebook_likes','actor_2_facebook_likes',
             'actor_3_facebook_likes','actor_2_name','cast_total_facebook_likes','actor_3_name',
            'duration','facenumber_in_poster','content_rating','country','movie_imdb_link','aspect_ratio',
            'plot_keywords'],axis=1)

director_name  num_critic_for_reviews        gross  \
0         James Cameron                   723.0  760505847.0   
1        Gore Verbinski                   302.0  309404152.0   
2            Sam Mendes                   602.0  200074175.0   
3     Christopher Nolan                   813.0  448130642.0   
4        Andrew Stanton                   462.0   73058679.0   
...                 ...                     ...          ...   
3848      Shane Carruth                   143.0     424760.0   
3849   Neill Dela Llana                    35.0      70071.0   
3850   Robert Rodriguez                    56.0    2040920.0   
3851       Edward Burns                    14.0       4584.0   
3852           Jon Gunn                    43.0      85222.0   

                                   genres     actor_1_name  \
0         Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1                Action|Adventure|Fantasy      Johnny Depp   
2               Action|Adventure|Thriller  Christoph Waltz   
3                         Action|Thriller        Tom Hardy   
4                 Action|Adventure|Sci-Fi     Daryl Sabara   
...                                   ...              ...   
3848                Drama|Sci-Fi|Thriller    Shane Carruth   
3849                             Thriller      Ian Gamazon   
3850  Action|Crime|Drama|Romance|Thriller  Carlos Gallardo   
3851                         Comedy|Drama      Kerry Bishé   
3852                          Documentary      John August   

                                    movie_title  num_voted_users  \
0                                       Avatar            886204   
1     Pirates of the Caribbean: At World's End            471220   
2                                      Spectre            275868   
3                        The Dark Knight Rises           1144337   
4                                  John Carter            212204   
...                                         ...              ...   
3848                                    Primer             72639   
3849                                    Cavite               589   
3850                               El Mariachi             52055   
3851                                 Newlyweds              1338   
3852                         My Date with Drew              4285   

      num_user_for_reviews language       budget  title_year  imdb_score  \
0                   3054.0  English  237000000.0      2009.0         7.9   
1                   1238.0  English  300000000.0      2007.0         7.1   
2                    994.0  English  245000000.0      2015.0         6.8   
3                   2701.0  English  250000000.0      2012.0         8.5   
4                    738.0  English  263700000.0      2012.0         6.6   
...                    ...      ...          ...         ...         ...   
3848                 371.0  English       7000.0      2004.0         7.0   
3849                  35.0  English       7000.0      2005.0         6.3   
3850                 130.0  Spanish       7000.0      1992.0         6.9   
3851                  14.0  English       9000.0      2011.0         6.4   
3852                  84.0  English       1100.0      2004.0         6.6   

      movie_facebook_likes  
0                    33000  
1                        0  
2                    85000  
3                   164000  
4                    24000  
...                    ...  
3848                 19000  
3849                    74  
3850                     0  
3851                   413  
3852                   456  

[3853 rows x 13 columns]

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'. 

In [9]:
# Write you code here
round(100*(movies.isnull().sum()/len(movies.index)), 2)

color                        0.05
director_name                0.00
num_critic_for_reviews       0.03
duration                     0.03
director_facebook_likes      0.00
actor_3_facebook_likes       0.16
actor_2_name                 0.03
actor_1_facebook_likes       0.00
gross                        0.00
genres                       0.00
actor_1_name                 0.00
movie_title                  0.00
num_voted_users              0.00
cast_total_facebook_likes    0.00
actor_3_name                 0.16
facenumber_in_poster         0.16
plot_keywords                0.78
movie_imdb_link              0.00
num_user_for_reviews         0.00
language                     0.08
country                      0.00
content_rating               1.25
budget                       0.00
title_year                   0.00
actor_2_facebook_likes       0.03
imdb_score                   0.00
aspect_ratio                 1.87
movie_facebook_likes         0.00
dtype: float64

#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [10]:
# Write your code for filling the NaN values in the 'language' column here
# Feedback:
# To replace the values, you can equate the entries with language column as null with 'English':

# movies.loc[pd.isnull(movies['language']), ['language']] = 'English'
 

# Next, you can count the movies made in English:

# (movies.language == 'English').sum()
movies.loc[pd.isnull(movies['language']), ['language']] = 'English'
(movies.language == 'English').sum()


3674

#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [13]:
# Write your code for unit conversion here
# Divide the 'gross' and 'budget' columns by 1000000 to convert '$' to 'million $'
movies.budget=movies.budget/1000000
movies.gross=movies.gross/1000000



In [14]:
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4  Color     Andrew Stanton                   462.0     132.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    475.0                   530.0   Samantha Morton   

   actor_1_facebook_likes       gross                           genres  ...  \
0                  1000.0  760.505847  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309.404152         Action|Adventure|Fantasy  ...   
2                 11000.0  200.074175        Action|Adventure|Thriller  ...   
3                 27000.0  448.130642                  Action|Thriller  ...   
4                   640.0   73.058679          Action|Adventure|Sci-Fi  ...   

  num_user_for_reviews language  country  content_rating    budget  \
0               3054.0  English      USA           PG-13  0.000237   
1               1238.0  English      USA           PG-13  0.000300   
2                994.0  English       UK           PG-13  0.000245   
3               2701.0  English      USA           PG-13  0.000250   
4                738.0  English      USA           PG-13  0.000264   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4      2012.0                  632.0        6.6          2.35   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                24000  

[5 rows x 28 columns]

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [15]:
# Write your code for creating the profit column here
movies['profit']=movies['gross']-movies['budget']

In [17]:
top_10 = movies.sort_values(by = 'profit', ascending = False).head(10)

In [19]:
# Write your code for sorting the dataframe here
movies.sort_values(by = 'profit', ascending = False).head(10)

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
25    Color      James Cameron                   315.0     194.0   
28    Color    Colin Trevorrow                   644.0     124.0   
16    Color        Joss Whedon                   703.0     173.0   
64    Color  Christopher Nolan                   645.0     152.0   
230   Color       George Lucas                   320.0     136.0   
2704  Color       George Lucas                   282.0     125.0   
7     Color        Joss Whedon                   635.0     141.0   
3     Color  Christopher Nolan                   813.0     164.0   
552   Color     Andrew Adamson                   205.0      93.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                         0.0                   855.0   Joel David Moore   
25                        0.0                   794.0       Kate Winslet   
28                      365.0                  1000.0         Judy Greer   
16                        0.0                 19000.0  Robert Downey Jr.   
64                    22000.0                 11000.0       Heath Ledger   
230                       0.0                  1000.0        Liam Neeson   
2704                      0.0                   504.0      Peter Cushing   
7                         0.0                 19000.0  Robert Downey Jr.   
3                     22000.0                 23000.0     Christian Bale   
552                      80.0                    48.0  Jennifer Saunders   

      actor_1_facebook_likes       gross  \
0                     1000.0  760.505847   
25                   29000.0  658.672302   
28                    3000.0  652.177271   
16                   26000.0  623.279547   
64                   23000.0  533.316061   
230                  20000.0  474.544677   
2704                 11000.0  460.935665   
7                    26000.0  458.991599   
3                    27000.0  448.130642   
552                    692.0  436.471036   

                                                 genres  ... language country  \
0                       Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
25                                        Drama|Romance  ...  English     USA   
28                     Action|Adventure|Sci-Fi|Thriller  ...  English     USA   
16                              Action|Adventure|Sci-Fi  ...  English     USA   
64                          Action|Crime|Drama|Thriller  ...  English     USA   
230                     Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
2704                    Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
7                               Action|Adventure|Sci-Fi  ...  English     USA   
3                                       Action|Thriller  ...  English     USA   
552   Adventure|Animation|Comedy|Family|Fantasy|Romance  ...  English     USA   

      content_rating    budget title_year  actor_2_facebook_likes imdb_score  \
0              PG-13  0.000237     2009.0                   936.0        7.9   
25             PG-13  0.000200     1997.0                 14000.0        7.7   
28             PG-13  0.000150     2015.0                  2000.0        7.0   
16             PG-13  0.000220     2012.0                 21000.0        8.1   
64             PG-13  0.000185     2008.0                 13000.0        9.0   
230               PG  0.000115     1999.0                 14000.0        6.5   
2704              PG  0.000011     1977.0                  1000.0        8.7   
7              PG-13  0.000250     2015.0                 21000.0        7.5   
3              PG-13  0.000250     2012.0                 23000.0        8.5   
552               PG  0.000150     2004.0                   309.0        7.2   

     aspect_ratio  movie_facebook_likes      profit  
0            1.78                 33000  760.505610  
25           2.35                 26000  658.672102  


In [21]:
top10 = movies.iloc[0:10]
top10

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4  Color     Andrew Stanton                   462.0     132.0   
5  Color          Sam Raimi                   392.0     156.0   
6  Color       Nathan Greno                   324.0     100.0   
7  Color        Joss Whedon                   635.0     141.0   
8  Color        David Yates                   375.0     153.0   
9  Color        Zack Snyder                   673.0     183.0   

   director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                      0.0                   855.0   Joel David Moore   
1                    563.0                  1000.0      Orlando Bloom   
2                      0.0                   161.0       Rory Kinnear   
3                  22000.0                 23000.0     Christian Bale   
4                    475.0                   530.0    Samantha Morton   
5                      0.0                  4000.0       James Franco   
6                     15.0                   284.0       Donna Murphy   
7                      0.0                 19000.0  Robert Downey Jr.   
8                    282.0                 10000.0   Daniel Radcliffe   
9                      0.0                  2000.0       Lauren Cohan   

   actor_1_facebook_likes       gross  \
0                  1000.0  760.505847   
1                 40000.0  309.404152   
2                 11000.0  200.074175   
3                 27000.0  448.130642   
4                   640.0   73.058679   
5                 24000.0  336.530303   
6                   799.0  200.807262   
7                 26000.0  458.991599   
8                 25000.0  301.956980   
9                 15000.0  330.249062   

                                              genres  ... language country  \
0                    Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
1                           Action|Adventure|Fantasy  ...  English     USA   
2                          Action|Adventure|Thriller  ...  English      UK   
3                                    Action|Thriller  ...  English     USA   
4                            Action|Adventure|Sci-Fi  ...  English     USA   
5                           Action|Adventure|Romance  ...  English     USA   
6  Adventure|Animation|Comedy|Family|Fantasy|Musi...  ...  English     USA   
7                            Action|Adventure|Sci-Fi  ...  English     USA   
8                   Adventure|Family|Fantasy|Mystery  ...  English      UK   
9                            Action|Adventure|Sci-Fi  ...  English     USA   

   content_rating    budget title_year  actor_2_facebook_likes imdb_score  \
0           PG-13  0.000237     2009.0                   936.0        7.9   
1           PG-13  0.000300     2007.0                  5000.0        7.1   
2           PG-13  0.000245     2015.0                   393.0        6.8   
3           PG-13  0.000250     2012.0                 23000.0        8.5   
4           PG-13  0.000264     2012.0                   632.0        6.6   
5           PG-13  0.000258     2007.0                 11000.0        6.2   
6              PG  0.000260     2010.0                   553.0        7.8   
7           PG-13  0.000250     2015.0                 21000.0        7.5   
8              PG  0.000250     2009.0                 11000.0        7.5   
9           PG-13  0.000250     2016.0                  4000.0        6.9   

  aspect_ratio  movie_facebook_likes      profit  
0         1.78                 33000  760.505610  
1         2.35                     0  309.403852  
2         2.35                 85000  200.073930  
3         2.35                164000  448.130392  
4         2.35                 24000   73.058415  
5         2.35    

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [38]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_250'

#IMDb_Top_250 = movies.sort_values(by = 'imdb_score', ascending = False).head(250)
#IMDb_Top_250['Rank']=range(1,251)
#IMDb_Top_250 = IMDb_Top_250.loc[IMDb_Top_250.num_voted_users > 25000]
IMDb_Top_250 = IMDb_Top_250.iloc[:250, ]
IMDb_Top_250



color         director_name  num_critic_for_reviews  duration  \
1795  Color        Frank Darabont                   199.0     142.0   
3016  Color  Francis Ford Coppola                   208.0     175.0   
2543  Color  Francis Ford Coppola                   149.0     220.0   
64    Color     Christopher Nolan                   645.0     152.0   
325   Color         Peter Jackson                   328.0     192.0   
...     ...                   ...                     ...       ...   
677   Color            Doug Liman                   249.0     119.0   
1734  Color          Pierre Morel                   309.0      93.0   
86    Color              Don Hall                   384.0     102.0   
746   Color         Cameron Crowe                   149.0     152.0   
157   Color          Dean DeBlois                   292.0     102.0   

      director_facebook_likes  actor_3_facebook_likes     actor_2_name  \
1795                      0.0                   461.0   Jeffrey DeMunn   
3016                      0.0                  3000.0    Marlon Brando   
2543                      0.0                  3000.0        Al Pacino   
64                    22000.0                 11000.0     Heath Ledger   
325                       0.0                   416.0       Billy Boyd   
...                       ...                     ...              ...   
677                     218.0                    73.0    Josh Hamilton   
1734                    180.0                   485.0    Holly Valance   
86                       38.0                   562.0    Daniel Henney   
746                     488.0                   947.0  Zooey Deschanel   
157                     255.0                   953.0   Djimon Hounsou   

      actor_1_facebook_likes       gross  \
1795                 11000.0   28.341469   
3016                 14000.0  134.821952   
2543                 22000.0   57.300000   
64                   23000.0  533.316061   
325                   5000.0  377.019252   
...                      ...         ...   
677                  13000.0  121.468960   
1734                 14000.0  145.000989   
86                     756.0  222.487711   
746                  22000.0   32.522352   
157                  18000.0  176.997107   

                                                 genres  ... language country  \
1795                                        Crime|Drama  ...  English     USA   
3016                                        Crime|Drama  ...  English     USA   
2543                                        Crime|Drama  ...  English     USA   
64                          Action|Crime|Drama|Thriller  ...  English     USA   
325                      Action|Adventure|Drama|Fantasy  ...  English     USA   
...                                                 ...  ...      ...     ...   
677                             Action|Mystery|Thriller  ...  English     USA   
1734                                    Action|Thriller  ...  English  France   
86    Action|Adventure|Animation|Comedy|Drama|Family...  ...  English     USA   
746                        Adventure|Comedy|Drama|Music  ...  English     USA   
157    Action|Adventure|Animation|Comedy|Family|Fantasy  ...  English     USA   

      content_rating    budget title_year  actor_2_facebook_likes imdb_score  \
1795               R  0.000025     1994.0                   745.0        9.3   
3016               R  0.000006     1972.0                 10000.0        9.2   
2543               R  0.000013     1974.0                 14000.0        9.0   
64             PG-13  0.000185     2008.0                 13000.0        9.0   
325            PG-13  0.000094     2003.0                   857.0        8.9   
...              ...       ...        ...                     ...        ...   
677            PG-13  0.000060     2002.0                   147.0        7.9   
1734           PG-13  0.000025     2008.0                   816.0        7.9   
86                PG  0.000165     2014.0 

#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

In [26]:
IMDb_Top_250[(IMDb_Top_250['imdb_score']>=7.5)&(IMDb_Top_250['imdb_score']<8)].shape

(35, 29)

In [27]:
IMDb_Top_250[(IMDb_Top_250['imdb_score']>=8)&(IMDb_Top_250['imdb_score']<8.5)].shape

(168, 29)

In [28]:
IMDb_Top_250[(IMDb_Top_250['imdb_score']>=8.5)&(IMDb_Top_250['imdb_score']<9)].shape

(43, 29)

In [29]:
IMDb_Top_250[(IMDb_Top_250['imdb_score']>=9)&(IMDb_Top_250['imdb_score']<9.5)].shape

(4, 29)

In [30]:
IMDb_Top_250[(IMDb_Top_250['imdb_score']>=9.5)&(IMDb_Top_250['imdb_score']<10)].shape

(0, 29)

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [42]:
# Write your code for creating three new dataframes here
Meryl_Streep = movies[movies['actor_1_name'] =='Meryl Streep']
Meryl_Streep.head()

color   director_name  num_critic_for_reviews  duration  \
392   Color    Nancy Meyers                   187.0     120.0   
1038  Color   Curtis Hanson                    42.0     111.0   
1132  Color     Nora Ephron                   252.0     123.0   
1322  Color   David Frankel                   208.0     109.0   
1390  Color  Robert Redford                   227.0      92.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
392                     278.0                   963.0     Hunter Parrish   
1038                    161.0                   132.0    Glenn Morshower   
1132                      0.0                   923.0  Mary Lynn Rajskub   
1322                     64.0                   505.0      Anne Hathaway   
1390                      0.0                 10000.0         Tom Cruise   

      actor_1_facebook_likes       gross                           genres  \
392                  11000.0  112.703470             Comedy|Drama|Romance   
1038                 11000.0   46.815748  Action|Adventure|Crime|Thriller   
1132                 11000.0   94.125426          Biography|Drama|Romance   
1322                 11000.0  124.732962             Comedy|Drama|Romance   
1390                 11000.0   14.998070               Drama|Thriller|War   

      ... language country  content_rating    budget title_year  \
392   ...  English     USA               R  0.000085     2009.0   
1038  ...  English     USA           PG-13  0.000045     1994.0   
1132  ...  English     USA           PG-13  0.000040     2009.0   
1322  ...  English     USA           PG-13  0.000035     2006.0   
1390  ...  English     USA               R  0.000035     2007.0   

      actor_2_facebook_likes imdb_score aspect_ratio  movie_facebook_likes  \
392                   2000.0        6.6         1.85                     0   
1038                   894.0        6.3         2.35                     0   
1132                   935.0        7.0         1.85                 13000   
1322                 11000.0        6.8         2.35                     0   
1390                 10000.0        6.2         2.35                     0   

          profit  
392   112.703385  
1038   46.815703  
1132   94.125386  
1322  124.732927  
1390   14.998035  

[5 rows x 29 columns]

In [43]:
Leo_Caprio = movies[movies['actor_1_name'] =='Leonardo DiCaprio']

In [44]:
Brad_Pitt = movies[movies['actor_1_name'] =='Brad Pitt']

In [46]:
# Write your code for combining the three dataframes here
Combined = pd.concat([Meryl_Streep,Leo_Caprio,Brad_Pitt])

In [57]:
# Write your code for grouping the combined dataframe here
Combined_bygroup=Combined.groupby(by=(Combined['actor_1_name']))
Combined_bygroup['num_critic_for_reviews'].mean()


actor_1_name
Brad Pitt            245.000000
Leonardo DiCaprio    330.190476
Meryl Streep         181.454545
Name: num_critic_for_reviews, dtype: float64

In [56]:
# Write the code for finding the mean of critic reviews and audience reviews here
Combined_bygroup['num_user_for_reviews'].mean()


actor_1_name
Brad Pitt            742.352941
Leonardo DiCaprio    914.476190
Meryl Streep         297.181818
Name: num_user_for_reviews, dtype: float64

#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>